### PR TITLE
Add file support to the bulk CDK

### PR DESCRIPTION
**Pull request by benmoriceau**
## What
Add the file transfer support for the bulk CDK.

## How
It:
- Create a queue for the file message
- When a a DestinationFile message is recieved it is publish in the queue
- The ProcessFileTask consume the queue, the file is then provided to the FilePartAccumulator
- The FilePartAccumulator breaks the file into 10Mb parts which are put into a batch. The CDK is then responsible to push those batch to the destination.
- Because of a aws sdk issue, the parallelism is set to 3 for number of parallel batch processing for the file

## Review guide
1. `DestinationTaskLauncher.kt`
2. `InputConsumerTask.kt`
3. `ProcessFileTask.kt`
4. `FilePartAccumulator.kt`

## Can this PR be safely reverted and rolled back?

- [x] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/write/StreamLoader.kt [hunk 1]
```diff
@@ -63,4 +67,9 @@ interface BatchAccumulator {
         throw NotImplementedError(
             "processRecords must be implemented if createBatchAccumulator is overridden"
         )
+
+    suspend fun processFilePart(file: DestinationFile, index: Long): Unit =
```
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/write/StreamLoader.kt:71]: Can we make this a separate accumulator? Since it's already created by a different method call?

File: airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ObjectStorageFormattingWriter.kt [hunk 2]
```diff
@@ -36,9 +36,9 @@ import java.io.Closeable
 import java.io.OutputStream
 import org.apache.avro.Schema
 
-interface ObjectStorageFormattingWriter : Closeable {
-    fun accept(record: DestinationRecord)
-    fun flush()
+abstract class ObjectStorageFormattingWriter : Closeable {
```
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ObjectStorageFormattingWriter.kt:39]: What is the purpose here? Should we revert this?

Comment (benmoriceau) [reply airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ObjectStorageFormattingWriter.kt:39]: Hmmm, I thought that I clean this. It is not necessary

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/write/StreamLoader.kt (hunk 1 above)
Comment (benmoriceau) [reply airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/write/StreamLoader.kt:71]: Will do.

File: airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/write/object_storage/FilePartAccumulator.kt [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import io.airbyte.cdk.load.message.object_storage.LoadablePart
+import io.airbyte.cdk.load.write.BatchAccumulator
+import java.io.File
+import java.nio.file.Path
+
+@SuppressFBWarnings(
+    "NP_NONNULL_PARAM_VIOLATION",
+    justification = "state is guaranteed to be non-null by Kotlin's type system"
+)
+class FilePartAccumulator(
```
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/write/object_storage/FilePartAccumulator.kt:26]: Can we add a comment here flagging that this is nonstandard usage. Specifically, the accumulators shouldn't be publishing to queues or doing bookkeeping, it's just awkward to do the file work without more of a flow interface. (because it goes `1 Record -> Many Parts` instead of `Many Records >- 1 Part`)

Like really, this should be `FilePartGenerator` (from `StreamLoader::createFilePartGenerator`) with something like

```
class FilePartGenerator(
  file: DestinationFile
)  {
  fun parts(): Sequence<Part> { ... }
}
```

Comment (johnny-schmidt) [reply airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/write/object_storage/FilePartAccumulator.kt:26]: (But no change required, since you've already tested this. I just don't want people to follow this pattern for non-file stuff.)

File: airbyte-cdk/bulk/toolkits/load-object-storage/src/test/kotlin/io/airbyte/cdk/load/write/object_storage/ObjectStorageStreamLoaderTest.kt
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/toolkits/load-object-storage/src/test/kotlin/io/airbyte/cdk/load/write/object_storage/ObjectStorageStreamLoaderTest.kt:1]: Why was this file deleted?

File: airbyte-integrations/connectors/destination-s3-v2/metadata.yaml
Comment (johnny-schmidt) [inline airbyte-integrations/connectors/destination-s3-v2/metadata.yaml:1]: Revert this! This will release S3.

File: airbyte-integrations/connectors/destination-s3-v2/src/test-integration/kotlin/io/airbyte/integrations/destination/s3_v2/S3V2WriteTest.kt [hunk 4]
```diff
@@ -72,6 +72,11 @@ class S3V2WriteTestJsonUncompressed :
     override fun testBasicTypes() {
         super.testBasicTypes()
     }
+
```
Comment (johnny-schmidt) [inline airbyte-integrations/connectors/destination-s3-v2/src/test-integration/kotlin/io/airbyte/integrations/destination/s3_v2/S3V2WriteTest.kt:75]: Remove these?

**Review by johnny-schmidt — CHANGES_REQUESTED** [review]
Definitely revert the metadata.yaml changes merging,

and any of the other flagged changes that are unnecessary

File: airbyte-integrations/connectors/destination-s3-v2/metadata.yaml
Comment (benmoriceau) [reply airbyte-integrations/connectors/destination-s3-v2/metadata.yaml:1]: Done

File: airbyte-cdk/bulk/toolkits/load-object-storage/src/test/kotlin/io/airbyte/cdk/load/write/object_storage/ObjectStorageStreamLoaderTest.kt
Comment (benmoriceau) [reply airbyte-cdk/bulk/toolkits/load-object-storage/src/test/kotlin/io/airbyte/cdk/load/write/object_storage/ObjectStorageStreamLoaderTest.kt:1]: It was only testing the old processFile

File: airbyte-integrations/connectors/destination-s3-v2/src/test-integration/kotlin/io/airbyte/integrations/destination/s3_v2/S3V2WriteTest.kt (hunk 4 above)
Comment (benmoriceau) [reply airbyte-integrations/connectors/destination-s3-v2/src/test-integration/kotlin/io/airbyte/integrations/destination/s3_v2/S3V2WriteTest.kt:77]: I need it.

File: airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/write/object_storage/FilePartAccumulator.kt (hunk 3 above)
Comment (benmoriceau) [reply airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/write/object_storage/FilePartAccumulator.kt:26]: I added a comment explaining why we need this pattern

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/write/StreamLoader.kt (hunk 1 above)
Comment (benmoriceau) [reply airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/write/StreamLoader.kt:78]: Done

**Review by johnny-schmidt — APPROVED** [review]
LGTM! Sorry I took so long!
